### PR TITLE
raft: hide Campaign rules on applying all entries

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/6013.

@LK4D4 can you please give this a try?

Previously we have some rules about when there are configuration changes. It is actually a little bit hard to follow for applications. Now we try to handle it internally inside raft.

### CODE REVIEW

File: raft/rawnode.go [hunk 1]
```diff
@@ -103,9 +103,14 @@ func NewRawNode(config *Config, peers []Peer) (*RawNode, error) {
 			r.addNode(peer.ID)
 		}
 	}
+
 	// Set the initial hard and soft states after performing all initialization.
 	rn.prevSoftSt = r.softState()
-	rn.prevHardSt = r.hardState()
+	if lastIndex == 0 {
```
Comment (xiang90) [inline raft/rawnode.go]: @bdarnell I try to fix the rawnode impl here. We still need to persist the initial hard state if raft is node recovered from an existing state.

**Comment by bdarnell** [comment]
LGTM

**Comment by LK4D4** [comment]
@xiang90 sure, will try tomorrow.

**Comment by xiang90** [comment]
@LK4D4 Progress?

**Comment by LK4D4** [comment]
@xiang90 I updated to your PR commit and it works perfectly.

**Comment by xiang90** [comment]
@LK4D4 OK. Thanks for confirming this for me!

**Comment by xiang90** [comment]
@gyuho Do not back port this. We need to have this tested by functional testing cluster for a while.

**Comment by gyuho** [comment]
Ok won't. Thanks!

File: raft/raft.go [hunk 2]
```diff
@@ -575,6 +574,15 @@ func (r *raft) poll(id uint64, v bool) (granted int) {
 func (r *raft) Step(m pb.Message) error {
 	if m.Type == pb.MsgHup {
 		if r.state != StateLeader {
+			ents, err := r.raftLog.entries(r.raftLog.applied+1, r.raftLog.committed-r.raftLog.applied)
```
Comment (siddontang) [inline raft/raft.go]: Hi @xiang90

The entries second argument is `maxsize`, but here seems using log count (committed - applied), not size.

File: raft/raft.go [hunk 3]
```diff
@@ -575,6 +574,15 @@ func (r *raft) poll(id uint64, v bool) (granted int) {
...
+			if err != nil {
+				r.logger.Panicf("unexpected error getting uncommitted entries (%v)", err)
```
Comment (siddontang) [inline raft/raft.go]: seems we want to get un applied entires? @xiang90
